driver-core/src/main/java/com/datastax/driver/core/policies/ExtendedRetryPolicy.java

```diff
@@ -0,0 +1,66 @@
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core.policies;
+
+import com.datastax.driver.core.ConsistencyLevel;
+import com.datastax.driver.core.SocketOptions;
+import com.datastax.driver.core.Statement;
+
+/**
+ * A policy that defines a default behavior to adopt in the event of unexpected errors.
+ * <p>
+ * This interface exists only for backward compatibility reasons: its methods should really be
+ * defined by {@link RetryPolicy}, but adding it after the fact would break binary compatibility.
+ * It should be merged into {@link RetryPolicy} in the next major release.
+ * <p>
+ * All retry policies shipped with the driver implement this interface.
+ */
+public interface ExtendedRetryPolicy extends RetryPolicy {
+
+    /**
+     * Defines whether to retry and at which consistency level on an
+     * unexpected error.
+     * <p>
+     * This method might be invoked in the following situations:
+     * <ol>
+     *     <li>On a client timeout, while waiting for the server response
+     *     (see {@link SocketOptions#getReadTimeoutMillis()});</li>
+     *     <li>On a connection error (socket closed, etc.);</li>
+     *     <li>When the contacted host replies with an error, such as
+     *     {@code OVERLOADED}, {@code IS_BOOTSTRAPPING}, {@code SERVER_ERROR}, etc.</li>
+     *     <li>When the driver encounters an internal error.</li>
+     * </ol>
+     *
+     * Note that when this method is invoked, <em>the driver cannot guarantee that a mutation has
+     * been effectively applied server-side</em>.
+     * The parameter {@code mightHaveBeenApplied} can however be inspected to determine whether the mutation
+     * is <em>likely</em> to have been applied.
+     * When this flag is {@code false}, then it is certain that the query hasn't been applied,
+     * and it is always safe to retry it; when it is
+     * {@code true}, there is a non-negligible chance that the query has reached the coordinator
+     * and that it has been applied already; a retry should only be attempted if the query is known to be idempotent.
+     *
+     *
+     * @param statement the original query that failed.
+     * @param cl the original consistency level for the operation.
+     * @param nbRetry the number of retries already performed for this operation.
+     * @param mightHaveBeenApplied whether the mutation is likely to have been applied server-side.
+     * @return the retry decision. If {@code RetryDecision.RETHROW} is returned,
+     * a {@link com.datastax.driver.core.exceptions.DriverException DriverException} will be thrown for the operation.
+     */
+    RetryDecision onUnexpectedError(Statement statement, ConsistencyLevel cl, int nbRetry, boolean mightHaveBeenApplied);
```

Comment:
I agree that this is currently too vague:

I can see three different error types i may want to handle differently:

* Client timeouts - I may just accept the fact that I have occasionally slow queries and treat every incidence of timeout the same or do it on a per statement basis.
* Connection/Transport errors - exceptions thrown while attempting the write/waiting for response.   I think the general case I'd always want to retry (unless non-idempotent)
* Server errors - Errors returned from the server (Protocol, Bootstrapping, Server error).  These are typically exceptional cases.

I think in all cases but the third, there is no need to get any more specific about the exception.  I'd propose the following:

```java
onClientTimeout(Statement statement, ConsistencyLevel cl, int nbRetry);
onConnectionError(Statement statement, ConsistencyLevel cl, int nbRetry, boolean mightHaveBeenApplied);
onServerError(Statement statement, ConsistencyLevel cl, int nbRetry, DriverInternalError exception)
```

Thoughts on this approach:

**Pros**:

* Clear (enough) definition of what the error is about without giving too much information (like exposing internal exceptions).
* It is easy for me as a user to understand why the method is called and what to do based on the method called.

**Cons**:

* The user has to implement three separate methods instead of one.  
* There may be cases where neither of these three methods are covered maybe?